File: android/src/main/java/io/parity/signer/screens/scan/transaction/components/TransactionSummaryModels.kt
package io.parity.signer.screens.scan.transaction.components

import io.parity.signer.components.sharedcomponents.KeyCardModelBase
import io.parity.signer.components.toImageContent
import io.parity.signer.domain.BASE58_STYLE_ABBREVIATE
import io.parity.signer.domain.abbreviateString
import io.parity.signer.ui.helpers.PreviewData
import io.parity.signer.uniffi.Address
import io.parity.signer.uniffi.Card
import io.parity.signer.uniffi.MTransaction

/**
 * Local version of [Mtransaction] for the case of signing elements
 */

data class SigningTransactionModel(
	val summaryModels: List<TransactionSummaryModel>,
	val keyModel: KeyCardModelBase?,
) {
	companion object {
		fun createStub(): SigningTransactionModel =
			SigningTransactionModel(
				summaryModels = listOf(
					TransactionSummaryModel(
						pallet = "Balances",
						method = "transfer_keep_alive",
						destination = "1219xC79CXV31543DDXoQMjuA",
						value = "0.2 WND",
						mTransactionIndex = 0,
					),
					TransactionSummaryModel(
						pallet = "Balances2",
						method = "transfer_keep_alive2",
						destination = "1219xC79CXV31543DDXoQMjuA2",
						value = "0.3 WND",
						mTransactionIndex = 1,
					)
				),
				keyModel = KeyCardModelBase(
					path = "//polkadot//1",
					seedName = "Parity Keys",
					base58 = "1219xC79CXV31543DDXoQMjuA",
					identIcon = PreviewData.exampleIdenticonPng,
					hasPassword = true,
					networkLogo = "kusama"
				)
			)
	}
}

fun List<IndexedValue<MTransaction>>.toSigningTransactionModels(): List<SigningTransactionModel> {
	val fullModelsList: List<SigningTransactionModel> =
		map { it.toSigningTransactionModel() }
	val signatures = fullModelsList.map { it.keyModel }.toSet()
	return signatures.map { signature ->
		SigningTransactionModel(keyModel = signature,
			summaryModels = fullModelsList
				.filter { it.keyModel == signature }
				.map { it.summaryModels }
				.flatten())
	}
}

private fun IndexedValue<MTransaction>.toSigningTransactionModel(): SigningTransactionModel {
	var pallet: String = ""
	var method: String = ""
	var destination: String = ""
	var parameter: String = ""

	val methodCards = value.content.method?.map { it.card } ?: emptyList()
	for (methodCard in methodCards) {
		when (methodCard) {
			is Card.PalletCard -> {
				pallet = methodCard.f
			}
			is Card.CallCard -> {
				method = methodCard.f.methodName
			}
			is Card.IdCard -> {
				destination = methodCard.f.base58.abbreviateString(
					BASE58_STYLE_ABBREVIATE
				)
			}
			is Card.BalanceCard -> {
				parameter = "${methodCard.f.amount} ${methodCard.f.units}"
			}
			else -> {
				//ignore the rest of the cards
			}
		}
	}
	return SigningTransactionModel(
		summaryModels = listOf(
			TransactionSummaryModel(
				pallet = pallet,
				method = method,
				destination = destination,
				value = parameter,
				mTransactionIndex = index
			)
		),
		keyModel = value.authorInfo?.let { author ->
			KeyCardModelBase(
				path = author.address.path,
				seedName = author.address.seedName,
				base58 = author.base58,
				networkLogo = value.networkInfo?.networkLogo,
				identIcon = author.address.identicon.toImageContent(),//.svgPayload, on iOS
				hasPassword = author.address.hasPwd
			)
		}
	)
}


data class TransactionSummaryModel(
	val pallet: String,
	val method: String,
	val destination: String,
	val value: String,
	val mTransactionIndex: Int
)

File: ios/PolkadotVault/Screens/Scan/Models/TransactionSummaryModels.swift
//
//  TransactionSummaryModels.swift
//  Polkadot Vault
//
//  Created by Krzysztof Rodak on 20/11/2022.
//

import Foundation
import SwiftUI

extension TransactionPreviewRenderable {
    init(_ transaction: MTransaction) {
        var pallet = ""
        var method = ""
        var destination = ""
        var value = ""
        let methodCards = transaction.content.method?.map(\.card) ?? []

        for methodCard in methodCards {
            if case let .palletCard(value) = methodCard {
                pallet = value
            }
            if case let .callCard(value) = methodCard {
                method = value.methodName
            }
            if case let .idCard(value) = methodCard {
                destination = value.base58.truncateMiddle()
            }
            if case let .balanceCard(cardValue) = methodCard {
                value = [cardValue.amount, cardValue.units].joined(separator: " ")
            }
        }
        summary = .init(
            pallet: pallet,
            method: method,
            destination: destination,
            value: value
        )
        if let author = transaction.authorInfo {
            signature = .init(
                path: author.address.displayablePath,
                name: author.address.seedName,
                network: transaction.networkInfo?.networkLogo,
                base58: author.base58,
                identicon: author.address.identicon,
                hasPassword: author.address.hasPwd
            )
        } else {
            signature = nil
        }
    }
}

struct TransactionDetailsRow: Equatable, Identifiable {
    let id = UUID()
    let key: String
    let value: String
}

struct TransactionPreviewRenderable: Equatable {
    let summary: TransactionSummaryModel
    let signature: TransactionSignatureRenderable?
}

struct TransactionSummaryModel: Equatable {
    let pallet: String
    let method: String
    let destination: String
    let value: String

    init(_: MTransaction) {
        pallet = ""
        method = ""
        destination = ""
        value = ""
    }

    init(
        pallet: String = "",
        method: String = "",
        destination: String = "",
        value: String = ""
    ) {
        self.pallet = pallet
        self.method = method
        self.destination = destination
        self.value = value
    }

    var asRenderable: [TransactionDetailsRow] {
        let labelKey = Localizable.TransactionSign.Label.Details.self
        return [.init(key: labelKey.pallet.string, value: pallet),
                .init(key: labelKey.method.string, value: method),
                .init(key: labelKey.destination.string, value: destination),
                .init(key: labelKey.value.string, value: value)]
    }
}

struct TransactionSignatureRenderable: Equatable {
    let path: String
    let name: String
    let network: String?
    let base58: String
    let identicon: Identicon
    let hasPassword: Bool

    init(
        path: String = "",
        name: String = "",
        network: String? = nil,
        base58: String = "",
        identicon: Identicon,
        hasPassword: Bool = false
    ) {
        self.path = path
        self.name = name
        self.network = network
        self.base58 = base58
        self.identicon = identicon
        self.hasPassword = hasPassword
    }
}
